### DISCUSSION TITLE
Passing point number to ComputeResidual Method

### DISCUSSION THREAD

**Discussion by Gaurav11ME** [su2code/SU2#1642] [General]
I am solving a problem where I need to add a source term to each point in the X-Momentum Equation. I have completed the following.
1. I have created a configuration option to check whether a source term file is present.
2. I have added a configuration option to read the source file.
3. I have added a function by modifying flow_sources.hpp and flow_sources.cpp for reading the source term values in a su2double type vector.
4.  I have added a method in CEulerSolver.cpp to get the residual for each point in the loop. In each iteration of the loop iPoint Variable stores the point number, but I could not find a way to pass the Point variable to the ComputeResidual in flow_sources.cpp.

Now, I need to add these values point by point to the X-Momentum Equation. I am adding a function where I am setting all other elements of the residual to 0.0, except for residual[1]. Is there a way to pass the Point number to the ComputeResidual Method? I can then pass this Point number to get the particular element of the Momentum_source vector. I checked its implementation in CNumerics.hpp, and the only input it accepts is a Cconfig type object.

**Comment by pcarruscag** [comment]
We have developer meetings Wednesdays at 3pm UK https://meet.jit.si/SU2_DevMeeting
It will be easier to help you with this there.

**Reply by Gaurav11ME** [reply]
Sure. I will attend it. Thanks.

**Reply by Gaurav11ME** [reply]
Hi,
I have made some modifications in the source code, by doing some research and as suggested in the developer's meeting. Regarding the loop over all the points in the domain in CIncEulerSolver.cpp, should I get the global point index?

My understanding is that the loop is running over all the volume and "iPoint" refers to the index of the volume in the ongoing loop iteration, so distribution over the processor will still refer to the right index. Is it correct?

Also, is it possible to use the new source-term values in the adjoint solver? I need to optimize the source-term values at each grid point using the following procedure.

1. Guess source values at grid points.
2. Run a RANS solution
3. Compare the RANS velocity field with the measured field
4. Use loss function = (u_rans-u_measure)^2
5. Use adjoint optimization to get the updated source field
6. Repeat 2-5 until convergence (Loss<set_threshold)

[Changes made in su2 source-code.pdf](https://github.com/su2code/SU2/files/8807139/Changes.made.in.su2.source-code.pdf)
